File: OJ/leetcode/3-Longest Substring Without Repeating Characters(不重复最长子串)/LongestSubstring.java

/**
 * Source : https://leetcode.com/problems/longest-substring-without-repeating-characters
 * Author : cloudy 
 * Date : 2018-7-23
 * 
 * Problem: 
Given a string, find the length of the longest substring without repeating characters.

Examples:
Given "abcabcbb", the answer is "abc", which the length is 3.
Given "bbbbb", the answer is "b", with the length of 1.
Given "pwwkew", the answer is "wke", with the length of 3. 
Note that the answer must be a substring, "pwke" is a subsequence and not a substring.

 * solution:
使用一个 HashSet记录出现过的字符，每次遍历时查找这个set是否存在，
若存在，则证明出现重复了，更新此时的长度，开启下一位的连续查询
 */

import java.util.HashSet;

public class LongestSubstring {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int len = s.length();
        int rst = 0;
        for (int i = 0; i < len; i++) {
            for (int j = i; j < len; j++) {
                Character c = s.charAt(j);
                if (set.contains(c)) {
                    break;
                } else {
                    set.add(c);
                }
            }
            rst = Math.max(maxLen, set.size());
            set.clear();
        }
        return rst;
        // int rst = 0;
        // HashSet<Character> set = new HashSet<>();
        // for (int i = 0; i < s.length(); i++) {
        //     int j = i;
        //     while (j < s.length() && !set.contains(s.charAt(j))) {
        //         set.add(s.charAt(j));
        //         j++;
        //     }
        //     rst = Math.max(rst, j - i);
        //     set.clear();
        // }
        // return rst;
    }

    public static void main(String[] args) {
        LongestSubstring solution = new LongestSubstring();
        String[] data = new String[] { "abcabcbb", "bbbbb", "pwwkew" ,"c"};
        for (String s : data) {
            System.out.println(s + "-->" + solution.lengthOfLongestSubstring(s));
        }
    }
}

File: OJ/leetcode/3-Longest Substring Without Repeating Characters/LongestSubstring.cpp
#include<iostream>
#include<string>
#include<map>
#include<vector>
using namespace std;

int lengthOfLongestSubstring(string s) {
    int rst = 0;
    map<char, int> recordMap;
    int j = 0;
    string rstStr;
    for(int i = 0; i < s.size(); ++i) {
        if(recordMap.count(s[i])) { //map.count() 0»ò1£¬map.find()µü´úÆ÷
            j = max(j, recordMap[s[i]] + 1);
        }
        recordMap[s[i]] = i;
        if(i - j + 1 > rst) {
            rst = i - j + 1;
            rstStr = s.substr(j, rst);
        }
    }
    cout << rstStr << " ";
    return rst;
}

int main() {
    vector<string> v;
    v.push_back("abcabcbb");
    v.push_back("bbbb");
    v.push_back("pwwkew");
    v.push_back("c");
    v.push_back("tmmzuxt");

    for(vector<string>::size_type i = 0; i < v.size(); i++) {
        cout << v[i] << "-- > " << lengthOfLongestSubstring(v[i]) << endl;;
    }
    return 0;
}
